packages/truffle-compile/compilerSupplier.js

```diff
@@ -230,42 +227,61 @@ CompilerSupplier.prototype.getVersionUrlSegment = function(
     }
   }
 
+  const versionToUse = findNewestValidVersion(version, allVersions);
+  if (versionToUse) return allVersions.releases[versionToUse];
+
   return null;
 };
 
+const findNewestValidVersion = (version, allVersions) => {
+  if (!semver.validRange(version)) return null;
+  let satisfyingVersions = Object.keys(allVersions.releases).map(
+    solcVersion => {
+      if (semver.satisfies(solcVersion, version)) return solcVersion;
+    }
+  );
```

Comment:
We can use `const` instead of `let` here if you just tack the filter onto the end... can even get rid of the extra `if` in the filter

```suggestion
  ).filter(v => v);
```